### PR TITLE
[6766] Endpoint post trainees withdraw validation

### PR DESCRIPTION
**Pull request by defong**
### Context
Validation for withdrawing trainee

### Changes proposed in this pull request
Added validation for withdrawing trainee

### Guidance to review
Scene setting for validations

### Important business

- [ ] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [ ] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?
- [ ] Do we need to send any updates to DQT as part of the work in this PR?
- [ ] Does this PR need an ADR?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

**Comment by defong** [comment]
[Endpoint: POST /trainees/withdraw validation](https://trello.com/c/v8ubWBWn/6766-endpoint-post-trainees-withdraw-validation)

File: app/controllers/api/trainees/withdraws_controller.rb [hunk 1]
```diff
@@ -5,12 +5,10 @@ module Trainees
     class WithdrawsController < Api::BaseController
       def create
         if withdraw_allowed?
-          if errored_params.any?
-            render_parameter_invalid(parameter_keys: errored_params)
-          else
-            withdraw_trainee
-
+          if withdraw_trainee
             render(json: { data: TraineeSerializer.new(trainee).as_json }, status: :accepted)
```
Comment (stevehook) [inline app/controllers/api/trainees/withdraws_controller.rb:9]: Not part of this PR, but I'm wondering whether we should be using status 200 (OK) rather than 202 (accepted) here because accepted is only meant for async (I think).

File: app/models/api/withdrawal_attributes.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    def initialize(trainee:)
+      @trainee = trainee
+      super
+    end
+
+    def save!
+      if valid?
+        trainee.update!(attributes.symbolize_keys.except(:reasons).merge(withdrawal_reasons:))
+        trainee.withdraw!
```
Comment (stevehook) [inline app/models/api/withdrawal_attributes.rb:39]: What happens if the model throws a validation error?

File: app/models/api/withdrawal_attributes.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    ATTRIBUTES.each do |attr|
+      attribute attr
+    end
+
+    def initialize(trainee:)
+      @trainee = trainee
+      super
+    end
+
+    def save!
```
Comment (stevehook) [inline app/models/api/withdrawal_attributes.rb:36]: I am wondering whether this method should be extracted into a separate service class. That way `WithdrawAttributes` would be more like a plain data object.

**Review by stevehook — COMMENTED** [review]
I like way this is structured now - having `WithdrawAttribute` model the request attributes and validations. I'm not sure the `save` method should be in the same class (see inline comment) but open to persuasion!

File: app/models/api/withdrawal_attributes.rb (hunk 2 above)
Comment (defong) [reply app/models/api/withdrawal_attributes.rb:39]: A 500 error i guess, it was lifted from here
https://github.com/DFE-Digital/register-trainee-teachers/blob/adf92bee4cb628e25f1d27d21b2dabeada3f264b/app/forms/withdrawal/confirm_details_form.rb#L26-L35

File: app/controllers/api/trainees/withdraws_controller.rb (hunk 1 above)
Comment (defong) [reply app/controllers/api/trainees/withdraws_controller.rb:9]: Yeah i think 200 should really be fine.

File: app/models/api/withdrawal_attributes.rb (hunk 3 above)
Comment (defong) [reply app/models/api/withdrawal_attributes.rb:36]: hmmm so you want something like this?


```ruby

results = WithdrawService.call

if results.success?
  render(json: { data: results.as_json }, status: :accepted)
else
  render_validation_errors(errors: results.errors)
end

```

Is this the bigger picture you want to aim for?

Comment (stevehook) [reply app/models/api/withdrawal_attributes.rb:36]: Yes I like that better because it separates the two concerns here nicely. And it might be a more sensible pattern where there is more business logic for other endpoints. What do you reckon?

File: app/controllers/concerns/api/validations_and_error_handling.rb
Comment (AdColvin) [inline app/controllers/concerns/api/validations_and_error_handling.rb:1]: Is this concern really needed now? It's just including a module and calling a single method from it. Could you have the error response logic in here directly, or turn the ErrorResponse module into a concern?

File: app/models/api/withdrawal_attributes.rb (hunk 3 above)
Comment (AdColvin) [reply app/models/api/withdrawal_attributes.rb:36]: I agree, having these attributes objects as simple data containers makes sense to me.

**Review by AdColvin — COMMENTED** [review]
This all looks great 👍 Just a couple of minor comments.

File: app/models/api/withdrawal_attributes.rb (hunk 3 above)
Comment (AdColvin) [reply app/models/api/withdrawal_attributes.rb:36]: In most cases, I imagine you could just delegate the save method directly to the ApplicationRecord. Then any logic around the saving (like queuing jobs etc) could be triggered with callbacks. You could still use a service to do the work, but you wouldn't have to remember to use it in these cases.

**Review by stevehook — APPROVED** [review]
This all looks good to me.

**Review by AdColvin — APPROVED** [review]
LGTM 👍
